app/code/Magento/InventoryCatalog/Plugin/Catalog/Model/Product.php

```diff
@@ -0,0 +1,54 @@
+<?php
+declare(strict_types=1);
+
+namespace Magento\InventoryCatalog\Plugin\Catalog\Model;
+
+class Product
+{
+    /**
+     * @var \Magento\InventorySalesApi\Model\GetAssignedStockIdForWebsiteInterface
+     */
+    protected $getAssignedStockIdForWebsite;
+    /**
+     * @var \Magento\Store\Model\StoreManagerInterface
+     */
+    protected $storeManager;
+    /**
+     * @var \Magento\InventorySalesApi\Api\IsProductSalableInterface
+     */
+    protected $isProductSalable;
+
+    /**
+     * Product constructor.
+     * @param \Magento\InventorySalesApi\Model\GetAssignedStockIdForWebsiteInterface $getAssignedStockIdForWebsite
+     * @param \Magento\Store\Model\StoreManagerInterface $storeManager
+     * @param \Magento\InventorySalesApi\Api\IsProductSalableInterface $isProductSalable
+     */
+    public function __construct(
+        \Magento\InventorySalesApi\Model\GetAssignedStockIdForWebsiteInterface $getAssignedStockIdForWebsite,
+        \Magento\Store\Model\StoreManagerInterface $storeManager,
+        \Magento\InventorySalesApi\Api\IsProductSalableInterface $isProductSalable
+    ){
+        $this->getAssignedStockIdForWebsite = $getAssignedStockIdForWebsite;
+        $this->storeManager                 = $storeManager;
+        $this->isProductSalable             = $isProductSalable;
+    }
+    /**
+     * @param \Magento\Catalog\Model\Product $subject
+     * @param bool $result
+     */
+    public function afterIsSalable(
```

Comment:
I'd suggest to try making it `around` plugin, as in this case result of initial execution is still overwritten.

Reply:
In case if plugin is `around` we can also check before execution if we are in multi-stock mode and just let old implementation to `proceed`.

Reply:
Really concerning part of this implementation is performance, as `isSalable` is called a lot and we are adding more resource intensive logic to every call. We might consider some kind of in-object caching to mitigate that.